File: tests/r_parity/74_cic.py
"""StatsPAI Changes-in-Changes parity (Python side) -- Module 74.

Runs ``sp.cic`` on a two-period panel; the companion R script runs
``qte::CiC`` on the same dumped CSV.

``sp.cic`` splits the four (group x time) cells on ``t == 0`` / ``t == 1``,
while ``qte::CiC`` takes explicit ``t`` / ``tmin1`` values, so the CSV keeps
the raw 1/2 coding and this side derives ``post``.

Tolerance: rel < 1e-6 on the ATT and on each quantile treatment effect.
Standard errors are not compared -- the R call runs with ``se = FALSE`` and
StatsPAI's come from its own bootstrap.
"""

from __future__ import annotations

import warnings

import numpy as np
import pandas as pd

from _common import ParityRecord, dump_csv, write_results

MODULE = "74_cic"
PROBS = [round(0.1 * k, 1) for k in range(1, 10)]


def _panel() -> pd.DataFrame:
    rng = np.random.default_rng(17)
    n = 600
    treat = rng.integers(0, 2, n)
    ui = rng.normal(0, 1, n)
    rows = []
    for i in range(n):
        for t in (1, 2):
            y = ui[i] + 0.5 * t + 2.0 * (treat[i] * (t == 2)) + rng.normal(0, 1)
            rows.append({"id": i + 1, "t": t, "treat": int(treat[i]), "y": y})
    return pd.DataFrame(rows)


def main() -> None:
    import statspai as sp

    df = _panel()
    dump_csv(df, MODULE)

    work = df.copy()
    work["post"] = (work["t"] == 2).astype(int)
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        fit = sp.cic(
            work,
            y="y",
            group="treat",
            time="post",
            quantiles=PROBS,
            n_boot=1,
            seed=0,
        )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="cic_ATT",
            estimate=float(fit.estimate),
            se=None,
            n=int(len(df)),
        )
    ]
    detail = fit.detail.set_index("quantile")
    for p in PROBS:
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"qte_{round(p * 100):02d}",
                estimate=float(detail.loc[p, "qte"]),
                se=None,
                n=int(len(df)),
            )
        )

    write_results(MODULE, "py", rows, extra={"estimator": "cic"})


if __name__ == "__main__":
    main()

File: tests/r_parity/74_cic.R
# StatsPAI Changes-in-Changes parity (R side) -- Module 74.
#
# Reads data/74_cic.csv and runs qte::CiC, the R implementation of the
# Athey-Imbens (2006) estimator. Tolerance: rel < 1e-6 on the ATT and on
# each quantile treatment effect.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(qte)
})

MODULE <- "74_cic"

df <- read_csv_strict(MODULE)
probs <- seq(0.1, 0.9, 0.1)

fit <- qte::CiC(
  y ~ treat, t = 2, tmin1 = 1, tname = "t", data = df,
  panel = TRUE, idname = "id", se = FALSE, probs = probs
)

rows <- list(
  parity_row(
    module    = MODULE,
    statistic = "cic_ATT",
    estimate  = as.numeric(fit$ate),
    se        = NA_real_,
    n         = nrow(df)
  )
)
for (k in seq_along(probs)) {
  rows[[length(rows) + 1]] <- parity_row(
    module    = MODULE,
    statistic = sprintf("qte_%02d", round(probs[k] * 100)),
    estimate  = as.numeric(fit$qte[k]),
    se        = NA_real_,
    n         = nrow(df)
  )
}

write_results(MODULE, rows, extra = list(
  package = "qte",
  version = as.character(utils::packageVersion("qte"))
))
